cirq-core/cirq/protocols/decompose_protocol.py

```diff
@@ -128,15 +157,27 @@ class SupportsDecomposeWithQubits(Protocol):
     def _decompose_(self, qubits: Tuple['cirq.Qid', ...]) -> DecomposeResult:
         pass
 
+    def _decompose_with_context_(
+        self, qubits: Tuple['cirq.Qid', ...], *, context: Optional[DecompositionContext] = None
+    ) -> DecomposeResult:
+        pass
+
 
-def _try_op_decomposer(val: Any, decomposer: Optional[OpDecomposer]) -> DecomposeResult:
+def _try_op_decomposer(
+    val: Any, decomposer: Optional[OpDecomposer], *, context: Optional[DecompositionContext] = None
+) -> DecomposeResult:
     if decomposer is None or not isinstance(val, ops.Operation):
         return None
-    return decomposer(val)
+    if 'context' in inspect.signature(decomposer).parameters:
+        assert isinstance(decomposer, OpDecomposerWithContext)
```

Comment:
why not just check that the decomposer is an instance of `OpDecomposerWithContext` if we will abort if it's not?

Reply:
The `assert isinstance()` is only to make mypy happy. For `@runtime_checkable` types, mypy doesn't verify the exact signature of the function, only checks that the attributes exist.  See https://docs.python.org/3/library/typing.html#typing.runtime_checkable